File: autolink/clipboard/ios/src/NativeClipboardModule.m
#import "NativeClipboardModule.h"

#import <UIKit/UIPasteboard.h>

// The @LynxNativeModule annotation is what cocoapods-lynx-library scans for;
// it expands to a harmless ObjC forward declaration when compiled.
@LynxNativeModule("NativeClipboardModule")
@implementation NativeClipboardModule

+ (NSString *)name {
  return @"NativeClipboardModule";
}

+ (NSDictionary<NSString *, NSString *> *)methodLookup {
  return @{
    @"setString" : NSStringFromSelector(@selector(setString:callback:)),
    @"getString" : NSStringFromSelector(@selector(getString:)),
  };
}

- (void)setString:(NSString *)text callback:(LynxCallbackBlock)callback {
  dispatch_async(dispatch_get_main_queue(), ^{
    UIPasteboard.generalPasteboard.string = text;
    callback(@"");
  });
}

- (void)getString:(LynxCallbackBlock)callback {
  dispatch_async(dispatch_get_main_queue(), ^{
    callback(UIPasteboard.generalPasteboard.string ?: NSNull.null);
  });
}

@end

File: app/iosApp/iosApp/NativeModules/NativeClipboardModule.swift
import UIKit

@objcMembers
final class NativeClipboardModule: NSObject, LynxModule {
  static let name = "NativeClipboardModule"
  static let methodLookup = [
    "setString": "setString:callback:",
    "getString": "getString:",
  ]

  override init() {
    super.init()
  }

  @objc(initWithParam:)
  init(param: Any) {
    super.init()
  }

  func setString(_ text: String, callback: LynxCallbackBlock) {
    UIPasteboard.general.string = text
    callback("")
  }

  func getString(_ callback: LynxCallbackBlock) {
    callback(UIPasteboard.general.string ?? NSNull())
  }
}
